clients/src/main/java/org/apache/kafka/common/internals/Plugin.java

```diff
@@ -0,0 +1,78 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.internals;
+
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Monitorable;
+import org.apache.kafka.common.metrics.internals.PluginMetricsImpl;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class Plugin<T> implements Supplier<T>, AutoCloseable {
+
+    private final T instance;
+    private final Optional<PluginMetricsImpl> pluginMetrics;
+
+    private Plugin(T instance, PluginMetricsImpl pluginMetrics) {
+        this.instance = instance;
+        this.pluginMetrics = Optional.ofNullable(pluginMetrics);
+    }
+
+    public static <T> Plugin<T> wrapInstance(T instance, Metrics metrics, String key) {
+        return wrapInstance(instance, metrics, () -> tags(key, instance));
+    }
+
+    private static <T> Map<String, String> tags(String key, T instance) {
+        Map<String, String> tags = new LinkedHashMap<>();
+        tags.put("config", key);
+        tags.put("class", instance.getClass().getSimpleName());
+        return tags;
+    }
+
+    public static <T> List<Plugin<T>> wrapInstances(List<T> instances, Metrics metrics, String key) {
+        List<Plugin<T>> plugins = new ArrayList<>();
+        for (T instance : instances) {
+            plugins.add(wrapInstance(instance, metrics, key));
+        }
+        return plugins;
+    }
+
+    public static <T> Plugin<T> wrapInstance(T instance, Metrics metrics, Supplier<Map<String, String>> tagsSupplier) {
+        PluginMetricsImpl pluginMetrics = null;
+        if (instance instanceof Monitorable && metrics != null) {
+            pluginMetrics = new PluginMetricsImpl(metrics, tagsSupplier.get());
+            ((Monitorable) instance).withPluginMetrics(pluginMetrics);
+        }
+        return new Plugin<>(instance, pluginMetrics);
+    }
+
+    @Override
+    public T get() {
+        return instance;
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (pluginMetrics.isPresent()) pluginMetrics.get().close();
+        if (instance instanceof AutoCloseable) ((AutoCloseable) instance).close();
```

Comment:
So far the only plugins that support this feature are all AutoCloseable. When we'll add support to plugin that don't implement Closeable/AutoCloseable, we'll need to adjust this method. But I think for now it's fine.

Reply:
If someday we missed the "new added plugin without `AutoCloseable`", it might cause the resource leak. I think we should either update the KIP and say "we only support plugins are Closeable, AutoCloseable", or we fix the issue here, with reflection or something else.